File: java/dxa-module-flex-layout/src/main/java/com/sdl/dxa/modules/flexlayout/model/TitledEntity.java
package com.sdl.dxa.modules.flexlayout.model;

import com.sdl.webapp.common.api.model.EntityModel;
import com.sdl.webapp.common.exceptions.DxaException;

import java.beans.BeanInfo;
import java.beans.Introspector;
import java.beans.PropertyDescriptor;

/**
 * Titled Entity. Is an wrapper around DXA entity model to extract
 * an title of it to be used in a titled entity layout container (such as tabs, accordions etc).
 *
 * @author nic
 */
public class TitledEntity {

    private EntityModel entity;
    private String title;
    private String fieldName;

    static final String[] TITLE_FIELD_NAMES = {
            "headline",
            "heading",
            "title"
    };

    public TitledEntity(EntityModel entity) throws DxaException {
        this.entity = entity;
        extractTitle(entity);
    }

    public String getTitle() {
        return this.title;
    }

    public String getFieldName() {
        return fieldName;
    }

    public EntityModel getEntity() {
        return entity;
    }

    private void extractTitle(EntityModel entity) throws DxaException {
        try {
            BeanInfo beanInfo = Introspector.getBeanInfo(entity.getClass());
            for ( String fieldName : TITLE_FIELD_NAMES ) {
                for (PropertyDescriptor property : beanInfo.getPropertyDescriptors() ) {
                    if ( property.getName().equals(fieldName) && property.getPropertyType().isAssignableFrom(String.class) ) {
                        this.title = (String) property.getReadMethod().invoke(entity);
                        this.fieldName = fieldName;
                        return;
                    }
                }
            }
        }
        catch ( Exception e ) {
            throw new DxaException("Could not extract title from entity: " + entity.getId(), e);
        }
    }
}

File: dotnet/SDL.DXA.Modules.FlexLayout/Models/TitledEntity.cs
using Sdl.Web.Common.Models;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;

namespace SDL.DXA.Modules.FlexLayout.Models
{
    /// <summary>
    /// Titled Entity. Is an wrapper around DXA entity model to extract
    /// an title of it to be used in a titled entity layout container(such as tabs, accordions etc).
    /// </summary>
    public class TitledEntity
    {
        // TODO: Have the field names configurable
        static string[] TITLE_FIELD_NAMES = {
                    "Headline",
                    "Heading",
                    "Title"
        };

        public TitledEntity(EntityModel entity)
        {
            this.Entity = entity;
            this.ExtractTitle(entity);
        }

        public string Title
        {
            get; private set;
        }

        public string FieldName
        {
            get; private set;
        }

        public EntityModel Entity
        {
            get; private set;
        }

        protected void ExtractTitle(EntityModel entity)
        {
            var properties = entity.GetType().GetProperties();
            foreach ( var fieldName in TITLE_FIELD_NAMES )
            {
                var property = entity.GetType().GetProperty(fieldName);
                if ( property != null )
                {
                    this.Title = property.GetValue(entity).ToString();
                    this.FieldName = fieldName;
                    return;
                }
            }
        }
    }
}